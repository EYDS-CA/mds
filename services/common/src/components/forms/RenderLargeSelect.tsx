import React, { FC } from "react";
import { Select, Form } from "antd";
import { BaseInputProps } from "./BaseInput";
import { IOption } from "@mds/common/interfaces/common/option.interface";

/**
 * @component RenderLargeSelect - Ant Design `AutoComplete` component for redux-form -- being used instead of 'RenderSelect' for large data sets that require a limit.
 */
interface LargeSelectProps extends BaseInputProps {
  dataSource: any[];
  handleSearch?: (value) => void | Promise<void>;
  handleSelect?: (value, option) => void | Promise<void>;
  handleFocus?: () => void | Promise<void>;
  selectedOption?: IOption

}

const RenderLargeSelect: FC<LargeSelectProps> = ({
  handleFocus = () => { },
  handleSearch,
  handleSelect,
  dataSource = [],
  ...props
}) => {
  const selectedInData = dataSource.includes((opt) => props.selectedOption?.value === opt.value);
  const data = dataSource;
  if (props.selectedOption?.value && !selectedInData) {
    data.unshift(props.selectedOption)
  }
  return (
    <Form.Item
      required={props.required}
      label={props.label}
      validateStatus={
        props.meta.touched ? (props.meta.error && "error") || (props.meta.warning && "warning") : ""
      }
      help={
        props.meta.touched &&
        ((props.meta.error && <span>{props.meta.error}</span>) ||
          (props.meta.warning && <span>{props.meta.warning}</span>))
      }
    >
      <Select
        loading={props.loading}
        virtual={false}
        disabled={props.disabled}
        id={props.id}
        dropdownMatchSelectWidth
        showSearch
        style={{ width: "100%" }}
        defaultActiveFirstOption={false}
        placeholder={props.placeholder}
        notFoundContent="Not Found"
        options={data}
        filterOption={() => true}
        value={props.selectedOption?.value}
        onSearch={handleSearch}
        onSelect={handleSelect}
        onChange={props.input.onChange}
        onFocus={(event) => {
          handleFocus();
          props.input.onFocus(event);
        }}
      />
    </Form.Item>
  )
};

export default RenderLargeSelect;
