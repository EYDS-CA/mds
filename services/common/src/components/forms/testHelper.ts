export const inputMeta = {
    touched: false,
    error: false,
    warning: false,
    autofilled: false,
    asyncValidating: false,
    dirty: false,
    dispatch: jest.fn(),
    form: "FORM",
    initial: undefined,
    invalid: false,
    pristine: true,
    submitFailed: false,
    submitting: false,
    valid: true,
    visited: false,
};

export const errorInputMeta = {
    touched: true,
    error: "Error message",
    warning: false,
    autofilled: false,
    asyncValidating: false,
    dirty: true,
    dispatch: jest.fn(),
    form: "FORM",
    initial: undefined,
    invalid: true,
    pristine: true,
    submitFailed: false,
    submitting: false,
    valid: true,
    visited: true,
}

// needs a value and a name
export const inputProps = {
    onBlur: jest.fn(),
    onChange: jest.fn(),
    onDragStart: jest.fn(),
    onDrop: jest.fn(),
    onFocus: jest.fn()
}