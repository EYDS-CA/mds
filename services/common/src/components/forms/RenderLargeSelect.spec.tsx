import React from "react";
import { render } from "@testing-library/react";
import RenderLargeSelect from "./RenderLargeSelect";
import { errorInputMeta, inputProps } from "./testHelper";

describe("RenderLargeSelect", () => {
    it("renders properly without required (optional), failed validation + label", () => {
        const { container } = render(<RenderLargeSelect
            id="field"
            label="Mock Label"
            required
            input={{ value: "", name: "field", ...inputProps }}
            meta={errorInputMeta}
            dataSource={[
                { label: "Opt 1", value: "opt1" },
                { label: "Opt 2", value: "opt2" },
                { label: "Opt 3", value: "opt3" }
            ]}
            selectedOption={{ label: "Opt 4", value: "opt4" }}
        />);
        expect(container).toMatchSnapshot();
    })
})