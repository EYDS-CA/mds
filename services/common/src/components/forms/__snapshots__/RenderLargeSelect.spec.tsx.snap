// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderLargeSelect renders properly without required (optional), failed validation + label 1`] = `
<div>
  <div
    class="ant-form-item ant-form-item-with-help ant-form-item-has-error"
  >
    <div
      class="ant-row ant-form-item-row"
    >
      <div
        class="ant-col ant-form-item-label"
      >
        <label
          class="ant-form-item-required"
          title="Mock Label"
        >
          Mock Label
        </label>
      </div>
      <div
        class="ant-col ant-form-item-control"
      >
        <div
          class="ant-form-item-control-input"
        >
          <div
            class="ant-form-item-control-input-content"
          >
            <div
              class="ant-select ant-select-in-form-item ant-select-status-error ant-select-single ant-select-show-arrow ant-select-show-search"
              style="width: 100%;"
            >
              <div
                class="ant-select-selector"
              >
                <span
                  class="ant-select-selection-search"
                >
                  <input
                    aria-activedescendant="field_list_0"
                    aria-autocomplete="list"
                    aria-controls="field_list"
                    aria-haspopup="listbox"
                    aria-owns="field_list"
                    autocomplete="off"
                    class="ant-select-selection-search-input"
                    id="field"
                    role="combobox"
                    type="search"
                    value=""
                  />
                </span>
                <span
                  class="ant-select-selection-item"
                  title="Opt 4"
                >
                  Opt 4
                </span>
              </div>
              <span
                aria-hidden="true"
                class="ant-select-arrow"
                style="user-select: none;"
                unselectable="on"
              >
                <span
                  aria-label="down"
                  class="anticon anticon-down ant-select-suffix"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="down"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                    />
                  </svg>
                </span>
              </span>
            </div>
          </div>
        </div>
        <div
          style="display: flex; flex-wrap: nowrap;"
        >
          <div
            class="ant-form-item-explain ant-form-item-explain-connected"
            role="alert"
          >
            <div
              class="ant-form-item-explain-error"
            >
              <span>
                Error message
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
